testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -0,0 +1,907 @@
+  private MockServerListener serverListener = new MockServerListener();
+  private ArgumentCaptor<Status> statusCaptor = ArgumentCaptor.forClass(Status.class);
+  private ArgumentCaptor<Throwable> throwableCaptor = ArgumentCaptor.forClass(Throwable.class);
+  private ArgumentCaptor<Metadata> metadataCaptor = ArgumentCaptor.forClass(Metadata.class);
+  private ArgumentCaptor<InputStream> inputStreamCaptor
+      = ArgumentCaptor.forClass(InputStream.class);
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Before
+  public void setUp() {
+    server = newServer();
+    client = newClientTransport();
+  }
+
+  @After
+  public void tearDown() throws InterruptedException {
+    if (client != null) {
+      // TODO(ejona): swap to shutdownNow
+      client.shutdown();
+    }
+    if (serverTransport != null) {
+      // TODO(ejona): swap to shutdownNow
+      serverTransport.shutdown();
+    }
+    if (server != null) {
+      server.shutdown();
+      assertTrue(serverListener.waitForShutdown(1000, TimeUnit.MILLISECONDS));
+    }
+  }
+
+  // TODO(ejona):
+  //   multiple streams on same transport
+  //   multiple client transports to same server
+  //   halfClose to trigger flush (client and server)
+  //   flow control pushes back (client and server)
+  //   flow control provides precisely number of messages requested (client and server)
+  //   onReady called when buffer drained (on server and client)
+  //   test no start reentrancy (esp. during failure) (transport and call)
+  //   multiple requests/responses (verifying contents received)
+  //   server transport shutdown triggers client shutdown (via GOAWAY)
+  //   queued message InputStreams are closed on stream cancel
+  //     (and maybe exceptions handled)
+
+  @Test
+  public void serverNotListening() {
+    server = null;
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    verify(mockClientTransportListener, timeout(1000)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(statusCaptor.capture());
+    assertCodeEquals(Status.UNAVAILABLE, statusCaptor.getValue());
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    verify(mockClientTransportListener, never()).transportReady();
+  }
+
+  @Test
+  public void clientStartStop() throws Exception {
+    server.start(serverListener);
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    client.shutdown();
+    verify(mockClientTransportListener, timeout(1000)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(any(Status.class));
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    server.shutdown();
+    assertTrue(serverListener.waitForShutdown(1000, TimeUnit.MILLISECONDS));
+    server = null;
```

Comment:
At the time teardown didn't have the waitForShutdown. Yeah, I can remove it here.